### ISSUE TITLE
Error: write after end

### ISSUE THREAD

**Issue by HaoyangFan96** [iccicci/rotating-file-stream#25]
Node version: `8.11.3`
OS: 64 bit Windows 10

I am using` rotating-file-stream` and `morgan` in my application to log all HTTP requests into daily rotated file. I have followed closely with the way that the docs of `morgan` recommended to use RFS:
[https://github.com/expressjs/morgan#log-file-rotation](https://github.com/expressjs/morgan#log-file-rotation)

Code:
`app.js`

```
// define the log directory
let logDirectory = path.join(__dirname, 'log');

// ensure log directory exists
fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);

// create a rotating write stream
let writeStream = rfs((time) => {
    if (!time) {
        return 'access.log';
    }
    let t = moment(time);
    return `${t.format('MM-DD-YYYY')}.log`;
}, {
    interval: '1d', // daily rotate
    path: logDirectory,
    rotationTime: false,
});


// set the logger, logging every request to the console
app.use(morgan('common', {
    stream: writeStream,
}));
```

Error message:
```
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: write after end
    at writeAfterEnd (_stream_writable.js:236:12)
    at RotatingFileStream.Writable.write (_stream_writable.js:287:5)
    at Array.logRequest (C:\Haoyang\source_code\myapp\server\node_modules\morgan\index.js:130:14)
    at listener (C:\Haoyang\source_code\myapp\server\node_modules\on-finished\index.js:169:15)
    at onFinish (C:\Haoyang\source_code\myapp\server\node_modules\on-finished\index.js:100:5)
    at callback (C:\Haoyang\source_code\myapp\server\node_modules\ee-first\index.js:55:10)
    at ServerResponse.onevent (C:\Haoyang\source_code\myapp\server\node_modules\ee-first\index.js:93:5)
    at emitNone (events.js:111:20)
    at ServerResponse.emit (events.js:208:7)
    at onFinish (_http_outgoing.js:720:10)

```

My goal is quite simple: I just want to let `morgan` direct all its log into the rotating write stream created via RFS, with filename to be in `MM-DD-YYYY.log` format for each day. I do a check to return a fixed file name `access.log` for the case when the `time` parameter is `null`. However, the error still persists

**Comment by iccicci** [comment]
Hi @HaoyangFan96,

plz try to add following code just after `let writeStream` statement:

```
writeStream.on("error", console.log);
writeStream.on("warning", console.log);
```

The first error should help you finding the root cause of your problem.

Hope this helps,
iCC

**Comment by bryandbor** [comment]
I'm having this same issue. When I add the above-mentioned logging, I see the following:
```
{ Error: Too many destination file attempts
    at RotatingFileStream.exhausted (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:181:13)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:197:24)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/Users/off-duty/Code/donor-easy-client/node_modules/rotating-file-stream/compress.js:222:15)
  code: 'RFS-TOO-MANY',
  attempts:
   { '/Users/off-duty/Code/donor-easy-client/build/log/2018-07-29.log': 1000 } }
```

For the record, my code is nearly identical to the OP.

**Comment by iccicci** [comment]
Hi @bryandbor 

your _generator function_ takes only one parameter `time`, it lacks of `index`.
Please read carefully the **Note** of following paragraph: https://www.npmjs.com/package/rotating-file-stream#function-filenametime-index

This should solve your problem,
iCC

**Comment by bryandbor** [comment]
Hello @iccicci,

Since I am not using a size restriction on the files, it seems strange that `index` would be required in my generator signature. Moreover, adding `index` did not seem to affect the error generated.

I apologize to the OP, but I'm hoping my additions to your issue will help resolve your issue as well. Here is my code:
```
import fs from 'fs';
import path from 'path';
import rfs from 'rotating-file-stream';
import moment from 'moment';

export const LOG_DIRECTORY = 'log';
const logDirectory = path.join(__dirname, LOG_DIRECTORY);
// ensure log directory exists
fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);

export const DEFAULT_LOG_FILENAME = 'access.log';
export const fileNameGenerator = (time, index) => {
  if (!time) {
    return DEFAULT_LOG_FILENAME;
  }
  
  const t = moment(time);
  return `${t.format('YYYY-MM-DD')}.log`;
};

// create a rotating write stream
const logStream = rfs(fileNameGenerator, {
  interval: '10s', // rotate daily
  rotationTime: false,
  path: logDirectory
});

logStream.on("error", console.log);
logStream.on("warning", console.log);

export default logStream;
```

The above code still generates the following error:
```
Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_stream_writable.js:243:12)
    at RotatingFileStream.Writable.write (_stream_writable.js:291:5)
    at Array.logRequest (/Users/off-duty/Code/donor-easy-client/node_modules/morgan/index.js:130:14)
    at listener (/Users/off-duty/Code/donor-easy-client/node_modules/on-finished/index.js:169:15)
    at onFinish (/Users/off-duty/Code/donor-easy-client/node_modules/on-finished/index.js:100:5)
    at callback (/Users/off-duty/Code/donor-easy-client/node_modules/ee-first/index.js:55:10)
    at ServerResponse.onevent (/Users/off-duty/Code/donor-easy-client/node_modules/ee-first/index.js:93:5)
    at ServerResponse.emit (events.js:187:15)
    at onFinish (_http_outgoing.js:683:10)
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

**Comment by iccicci** [comment]
Hi @bryandbor ,

you are right! Sorry.

By the way, only adding `index` parameter would have not fixed the problem it should be used in the `return` statement as well... anyway (as you made me notice) this is not the case due you are not using file size limit.

The point is:
```
  return `${t.format('YYYY-MM-DD')}.log`;
```
and
```
  interval: '10s', // rotate daily
```
you are going to create 8640 times per day the same file. ;)

You need to change the _interval_ to `1d` or the _format_ to `YYY-MM-DD-HH-mm-ss`.

I'm pretty sure this time we reach the solution of your problem.

Thank you for using **rfs**,
iCC

**Comment by bryandbor** [comment]
Changing the filename to include hour, minute, and second did resolve the error being thrown; however, I had only lowered the interval to that level because an interval of `'1d'` would always call my `fileNameGenerator` function with a `time` value of `null` initially. 

After looking back over your API, I experimented by adding `immutable: true` to my options and this does provide a `time` value initially; however, the value seems to ignore my option for `rotationTime` as the value always represents the moment the code runs rather than having the ability to use the "start" of the interval. For my purposes, I can just truncate the `time` value to the start of the day, but for more detailed filenames, this may not be sufficient. I just wanted to bring this case to your attention.

---

There are two issues to point out here:
Sample filename generator:
```
export const DEFAULT_LOG_FILENAME = 'access.log';
export const fileNameGenerator = (time) => {
  console.log(time);
  if (!time) {
    return DEFAULT_LOG_FILENAME;
  }
  
  const t = moment(time);
  return `${t.format('YYYY-MM-DD-HH-mm-ss')}.log`;
};
```

1 If `immutable` is `undefined` or `false`, then `time` is `null` on the initial `fileNameGenerator` call regardless of the `rotationTime` option.
```
const stream1 = rfs(fileNameGenerator, {
  path: logDirectory,
  interval: '10s',
  rotationTime: true,
});
const stream2 = rfs(fileNameGenerator, {
  path: logDirectory,
  interval: '10s',
  rotationTime: false,
});
```
Produces these logs:
```
null
null
...
2018-07-30T15:54:30.000Z
2018-07-30T15:54:30.000Z
...
```
If `immutable` is not used, then the initial value of `time` is null, which means the initial logging file will always be an anti-pattern which seems incorrect.

---

2 If `immutable` is `true`, then `time` has a value upon initial call of `fileNameGenerator`; however, the `rotationTime` option is ignored on the initial call of `fileNameGenerator`; the exact time of calling is passed rather than passing the time of the "beginning" of the interval if `rotationTime` is set correctly.
```
const stream1 = rfs(fileNameGenerator, {
  path: logDirectory,
  interval: '10s',
  rotationTime: true,
  immutable: true,
});
const stream2 = rfs(fileNameGenerator, {
  path: logDirectory,
  interval: '10s',
  rotationTime: false,
  immutable: true,
});
```

The logs show:
```
2018-07-30T15:16:51.779Z
2018-07-30T15:16:51.779Z
...
2018-07-30T15:17:00.004Z
2018-07-30T15:17:00.005Z
...
```

In conclusion, I cannot understand why there is any situation in which `time` would not have a value if `interval` is set. If this is the intended behavior, then it seems strange since a larger interval option would always result in a default filename being used until the end of the initial interval **and** each time the server is required to restart. If I'm using this package, it seems to me that an `interval` or `size` should mean that there is never a situation where I need to use a "fallback" file name since I should always expect arguments which allow my function to generate a unique filename.

**Comment by iccicci** [comment]
Hi @JcBernack ,

everything you described respects the expected behaviour.
The reason why the _filename generator function_ is called with a **null** `time` parameter is to let the user to follow the log with `tail -F logfilename` .

The most probable reason why you get the WRITE AFTER END error is that the stream was closed due some previous error.

Pleas keep active the error and the warning event handlers
```
logStream.on("error", console.log);
logStream.on("warning", console.log);
```
the key should be in the very first logged error.

The options `rotationTime` and `immutable` abviously change the behaviour of the **rfs** but I'm pretty sure they have no impact on your error.

Hope this helps,
iCC

**Comment by HaoyangFan96** [comment]
@iccicci sorry for my late response. It seems like the error only occurs during the first midnight and later it works fine. Thank for your help and efforts in supporting us such a useful library 

If @bryandbor has no questions as well, then I'am fine with closing this issue

**Comment by ArtixZ** [comment]
I'm having the same error message, and here are the error logs:

```javascript
 Error: Too many destination file attempts
    at RotatingFileStream.exhausted (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:181:13)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:197:24)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
    at RotatingFileStream.findName (/home/ec2-user/app/code/server/node_modules/rotating-file-stream/compress.js:222:15)
  code: 'RFS-TOO-MANY',
  attempts: { '/home/ec2-user/logs/cmnportal-2018-10_26.log': 1000 } }
Error: write after end
    at writeAfterEnd (_stream_writable.js:236:12)
    at RotatingFileStream.Writable.write (_stream_writable.js:287:5)
    at Array.logRequest (/home/ec2-user/app/code/server/node_modules/morgan/index.js:130:14)
    at listener (/home/ec2-user/app/code/server/node_modules/on-finished/index.js:169:15)
    at onFinish (/home/ec2-user/app/code/server/node_modules/on-finished/index.js:100:5)
    at callback (/home/ec2-user/app/code/server/node_modules/ee-first/index.js:55:10)
    at ServerResponse.onevent (/home/ec2-user/app/code/server/node_modules/ee-first/index.js:93:5)
    at emitNone (events.js:111:20)
    at ServerResponse.emit (events.js:208:7)
    at onFinish (_http_outgoing.js:720:10)
```
Here are my code:
```javascript
let logDirectory = '/home/ec2-user/logs/'
fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory)
function pad(num) {
  return (num > 9 ? "" : "0") + num;
}
function generator(time, idx) {
  if(!!!time) {
    const now = new Date()
    return `cmnportal-${now.getFullYear()}-${pad(now.getMonth()+1)}_${now.getDate()}.log`
  }
  return `cmnportal-${time.getFullYear()}-${pad(time.getMonth()+1)}_${time.getDate()}.log`
}
const logStream = rfs(generator, {
  interval: '1m', // rotate daily
  path: logDirectory
})
```

**Comment by ArtixZ** [comment]
This config seems work for me:
```javascript
const logStream = rfs(generator, {
  interval: '1m', // rotate daily
  path: path.resolve(logDirectory),
  immutable: true,
})
```
@bryandbor : I believe `rotationTime: true` means `time` parameter will be fixed to the moment that you call the function( run the server up ).
